_posts/pl/2021-09-31-zarzadzanie-stanem-aplikacji-frontendowej.md

```diff
@@ -0,0 +1,225 @@
+```
+
+Każda akcja musi mieć zdefiniowany swój typ. NgRx używa do tego tzw. _literal type_, czyli po prostu używa typu string do identyfikacji akcji (u nas odpowiednio `Add activity` oraz `Get activity`. W nazywaniu akcji mamy pełną dowolność. Dodatkowo w przypadku dodawania nowej aktywności do listy musimy ją przekazać w ciele akcji przy pomocy metody `props<>()`. Dzięki wykorzystaniu funkcji `createAction` z biblioteki NgRx tworzenie akcji, jak widać, jest bardzo proste.
+
+### Efekt
+Jak wspomniałem, będziemy korzystać z zewnętrznego API. W tym celu napiszemy efekt, który będzie reagował na akcję `Get activity`, następnie wystosowywał żądanie HTTP, a po otrzymaniu odpowiedzi wywoływał akcję `Add Activity`.
+
+```typescript
+@Injectable()
+export class ActivityEffect {
+    constructor(private actions$: Actions,
+                private http: HttpClient) {
+    }
+
+  getActivity$ = createEffect(() => this.actions$.pipe(
+    ofType(getActivityType),
+    mergeMap(() => this.http.get<ActivityItemModel>('https://www.boredapi.com/api/activity')
+      .pipe(
+        map(response => (addActivity(response))),
+        catchError(() => EMPTY)
+      ))
+  ));
+}
+```
+
+Pojawia się tutaj dość dużo zagadnień, zatem spójrzmy na ten fragment kodu od góry do dołu. Najpierw informujemy Angulara, że efekt to tak naprawdę wstrzykiwalna klasa, podobnie jak zwykłe serwisy. W konstruktorze zapewniamy dostęp do serwisu HTTP oraz do zmiennej typu `Actions`. Jest to w gruncie rzeczy serwis udostępniający wszystkie wywołane akcje od momentu ostatniego uruchomienia reducera (czyli dzięki temu damy radę reagować na przychodzące akcje). 
+
+Następnie definiujemy sam efekt:
+* zaglądamy do strumienia akcji 
+```typescript 
+this.actions$.pipe()
+```
+* szukamy tylko akcji określonego typu 
+```typescript
+ofType(getActivityType)
+```
+* reagujemy na każdą z nich poprzez wywołanie żądania HTTP i zaglądamy do strumienia zawierającego odpowiedź z API 
+```typescript 
+mergeMap(() => this.http.get<ActivityItemModel>('https://www.boredapi.com/api/activity').pipe()
+```
+* mapujemy odpowiedź na wywołanie akcji `map(response => (addActivity(response)))`
+```typescript 
+map(response => (addActivity(response)))
+```
+* w przypadku błędu zwracamy pustą wartość Obervable
+```typescript 
+catchError(() => EMPTY)
+```
+
+W wyniku działania tego efektu każde wywołanie akcji `Get Activity` przy poprawnej odpowiedzi z API wywoła akcję `Add Activity` z otrzymaną czynnością.
+
+### Reducer
+W reducerach zazwyczaj znajduje się najwięcej logiki ze wszystkich komponentów NgRx. To tutaj trzeba zdecydować jak dodawać do store kolejne dane. Warto pamiętać, że reducery to **czyste funkcje**, które przyjmują jako parametr akcję i obecny stan aplikacji, a następnie zwracają nowy stan, dzięki czemu zachowujemy jego niemutowalność. W naszym przypadku wygląda to tak:
+
+```typescript
+const initialState: ActivityItemsState = {activities: []};
```

Comment:
pokażesz też unit test? bo to jest właśnie petarda w reduxie. że to co testujesz to reducery a testuje się je turbo łatwo bo to pure functions :man_dancing:

Reply:
Tak. Mówię o łatwości testowania, a nigdzie faktycznie tego nie pokazuję. Dodam.